pilot/pkg/model/gateway.go

```diff
@@ -231,10 +231,11 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
 							log.Warnf("TLS server without TLS options %s %s", gatewayName, s.String())
 							continue
 						}
-						if mergedServers[serverPort] == nil {
-							mergedServers[serverPort] = &MergedServers{Servers: []*networking.Server{s}}
-						} else {
-							mergedServers[serverPort].Servers = append(mergedServers[serverPort].Servers, s)
+						for port := range mergedServers {
+							if port.Number == serverPort.Number {
+								mergedServers[port].Servers = append(mergedServers[port].Servers, s)
```

Comment:
I am a bit concerned about this. In buildGatewayListeners, we use the key.Protocol field, which will now be wrong here, as with .Bind. Also, its n^2 and folks have O(10k) gateways so it may be a scalability issue